program/rust/src/rust_oracle.rs

```diff
@@ -23,5 +33,94 @@ pub fn update_version(
     instruction_data: &[u8],
 ) -> OracleResult {
     panic!("Need to merge fix to pythd in order to implement this");
-    Ok(0) //SUCCESS
+    Ok(SUCCESS)
 }
+
+
+/// initialize the first mapping account in a new linked-list of mapping accounts
+/// accounts[0] funding account           [signer writable]
+/// accounts[1] new mapping account       [signer writable]
+pub fn init_mapping(
+    program_id: &Pubkey,
+    accounts: &Vec<AccountInfo>,
+    instruction_data: &[u8],
+) -> OracleResult {
+    pyth_assert(accounts.len() == 2 &&
+                  valid_funding_account(accounts.get(0).unwrap()) &&
+                  valid_signable_account(program_id, accounts.get(1).unwrap(), size_of::<pc_map_table_t>()),
+                ProgramError::InvalidArgument)?;
+
+    // Check that the account has not already been initialized
+    let mapping_account = load_account_as::<pc_map_table_t>(&accounts.get(1).unwrap())
+      .map_err(|_| ProgramError::InvalidArgument)?;
+    pyth_assert(mapping_account.magic_ == 0 && mapping_account.ver_ == 0, ProgramError::InvalidArgument)?;
+
+    // Initialize by setting to zero again (just in case) and setting
+    // the version number
+    let hdr = load::<cmd_hdr_t>(instruction_data).map_err(|_| ProgramError::InvalidArgument)?;
+    let mut data = accounts.get(1).unwrap().try_borrow_mut_data().map_err(|_| ProgramError::InvalidArgument)?;
+    sol_memset( *data, 0, size_of::<pc_map_table_t>() );
+
+    let mut mapping_account = load_account_as_mut::<pc_map_table_t>(&accounts.get(1).unwrap())
+      .map_err(|_| ProgramError::InvalidArgument)?;
+    mapping_account.magic_ = PC_MAGIC_V;
+    mapping_account.ver_   = hdr.ver_;
+    mapping_account.type_  = PC_ACCTYPE_MAPPING_V;
+    mapping_account.size_  = (size_of::<pc_map_table_t>() - size_of_val( &mapping_account.prod_ )) as u32;
+
+    Result::Ok(SUCCESS)
+}
+
+// FIXME: this is an extremely scary way to check errors because if you forget the ? after calling it, it doesn't do anything.
+pub fn pyth_assert(condition: bool, error_code: ProgramError) -> Result<(), ProgramError> {
```

Comment:
thoughts on how to make this less error prone? I think we need some error-checking utilities, but this approach scares me.

Reply:
I think clippy catches this.

Reply:
I think if we turn it into a macro we can be sure it returns an error.
I am also looking into how we can add a clippy check that all errors have been propagated into out CI/CD.

Reply:
clippy gives you a warning, but i don't think that's good enough. I'll look into making a macro as a follow up pr.

Reply:
Clippy will catch it with the deny warnings option, which is on.